src/ConnectedMode/BranchMatcher.cs

```diff
@@ -0,0 +1,105 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2022 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using LibGit2Sharp;
+using SonarQube.Client;
+
+namespace SonarLint.VisualStudio.ConnectedMode
+{
+    public interface IBranchMatcher
+    {
+        Task<string> GetMatchedBranch(string projectKey);
```

Comment:
Needs a doc comment

Reply:
The doc comment is useful. I actually meant that we should explicitly state that we are using the same algorithm as the other flavours, since it isn't clear. Also, my understanding from our offline discussion was that we know that our implementation is different in some respects. It would be useful to record that while we still remember what the difference is.

Maybe add something like the following (assuming it's accurate):

> This is the same algorithm as the other SonarLint flavours, except in how we treated branch histories.
We treat branch histories as series of linear commits ordered by time, even if there are merges.